shell/platform/darwin/ios/framework/Source/FlutterChannels.mm

```diff
@@ -32,14 +32,14 @@ - (instancetype)initWithName:(NSString*)name
   self = [super init];
   NSAssert(self, @"Super init cannot be nil");
   _name = [name retain];
-  _messenger = [messenger retain];
+  _messenger = messenger;
```

Comment:
Here and elsewhere: what do we do if someone has a referenc eto a message channel and is sending it messages, but there's no messenger anymore?  Log a warning?  assert?

Reply:
to breake the circular reference we should not  retain messager， because the messager is holded by channel‘s caller

Reply:
Yes, but what happens if a caller with a reference to the channel tries to use it? This fixes the retain cycle but we could end up with zombies here, right?

Reply:
the messager actually is flutterengine which is holded by flutterviewcontroller or created by user, if flutterviewcontroller is existed, the engine existed, could  not be zombies

Reply:
But someone might have grabbed a reference to a message channel from a flutter engine or viewcontroller, and then try to use it after the engine is gone.

Reply:
In your case, when engine is gone, the engine is nil,  nil call any method and would not work, may not be crashed. 
this is my demo https://github.com/Natoto/flutterOnExistApp/tree/multiflutter
then you can see the total memory released,when close flutter viewcontroller
can you show your case when case crash.

Reply:
> This code will result in a crash based on the current version of the patch:
> 
> ```objc
>   FlutterBasicMessageChannel* channel;
>   @autoreleasepool {
>     FlutterViewController* flutterViewController =
>       [[FlutterViewController alloc] init];
>     channel = flutterViewController.engine.systemChannel;
>   }
>   [channel sendMessage:@"Hello!"];
> ```
> 
> You could imagine other steps in there, such as showing and then dismissing the FlutterViewController. We should not crash in this instance.

a compromise solution in FlutterEngine.m  method
 
avoid access zombies object and crash ,_shell maybe not reset succ or engine may be release but not nil ,use lifecyclechannel check if engine is gone
```
- (void)sendOnChannel:(NSString*)channel
              message:(NSData*)message
          binaryReply:(FlutterBinaryReply)callback{
 ...

    if(_lifecycleChannel.get()){
        _shell->GetPlatformView()->DispatchPlatformMessage(platformMessage);
    }
```

Reply:
If we don't reset the shell the engine can't be released :(